### ISSUE TITLE
FormatOutput and GAP strings

### ISSUE THREAD

**Issue by ChrisJefferson** [gap-system/gap#2317]
`FormatOutput` is a GAP kernel function which does low-level printing. It understands C strings for example, by writing `%s`. This function is used in many error-related functions. For example, in FuncOUTPUT_LOG_TO, streams.c

```
   ErrorReturnVoid( "OutputLogTo: cannot log to %s",
                         (Int)CSTR_STRING(filename), 0L,
                         "you can 'return;'" );
```

Or gvars.c, AssGVar:

```
ErrorReturnVoid(
            "Variable: '%s' is read only",
            (Int)NameGVar(gvar), 0L,
            "you can 'return;' after making it writable" );
```

Both of these pieces of code are invalid, because before `FormatOutput` is eventually called inside `ErrorReturnVoid`, an object is allocated, possibly causing a GC which will make the `CSTR_STRING` in the first case, and the `NameGVar` in the second case, invalid.

There are similar problems in a few other functions, and it's very easy to introduce similar bugs.

I can see two fixes for this:

1) Extend `FormatOutput` to understand GAP strings directly (we could require it only understand strings in `StringRep`. This would be quickest.

2) Add and make use of some slightly higher-level formatting function, which only uses GAP-level strings.

**Comment by fingolfin** [comment]
IIRC there are already several places which copy strings into stack buffers for exactly that reason; IIRC Code for printing large integers is affected... need to
look agsin later.

**Comment by ChrisJefferson** [comment]
Copying into a stack buffer would indeed be another option.

**Comment by ChrisJefferson** [comment]
I have put a possible fix for this in #2318
